man/xoreos.6
.Dd July 24, 2015
.Dt XOREOS 6
.Os
.Sh NAME
.Nm xoreos
.Nd reimplementation of BioWare's Aurora engine
.Sh SYNOPSIS
xoreos
.Op Ar options
.Op Ar target
.Sh DESCRIPTION
.Nm xoreos
is an open source implementation of BioWare's Aurora engine and its
derivatives, licensed under the terms of the GNU General Public
License version 3 (or later).
The goal is to have all games using this engine working in a portable manner.
.Pp
.Sy The following games are valid targets for xoreos:
.Bl -bullet -compact
.It
Neverwinter Nights
.It
Neverwinter Nights 2
.It
Knights of the Old Republic
.It
Knights of the Old Republic II: The Sith Lords
.It
Jade Empire
.It
Sonic Chronicles: The Dark Brotherhood
.It
The Witcher
.It
Dragon Age: Origins
.It
Dragon Age II
.El
.Ss Running xoreos
First, you need to fully install and/or copy the game you want to play
with
.Nm
onto your hard disk.
How you do this depends on the game,
your operating system and where/how you have bought the game.
.Pp
.Nm
does not yet have a launcher GUI or anything like this.
You need to start it from the command line.
Run xoreos with the command line option
.Fl Fl help
to get a help text about further command line options.
.Pp
The quickest way to start a game in path
.Pa /path/to/game/
would be to call:
.Pp
.Dl $ xoreos -p/path/to/game/
.Pp
If you're on Windows and the path is, say,
.Pa D:\ePath\eTo\eGame\e , call :
.Pp
.Dl $ xoreos -pD:\ePath\eTo\eGame\e
.Ss Config file
In general,
.Nm
can read the configuration which game to run from either the command
line, a config file or both.
Additionally, when you first specify a new game on the command line,
.Nm
will add a related entry in the config file (creating it first, if
necessary).
.Pp
To accurately identify a specific instance of an installed game,
.Nm
uses the concept of a
.Dq target .
Each target has a separate section in the config file,
and each of their options apply only to that target.
The special target
.Dq xoreos
is a global section applying to all games,
although the same option in a game target overrides the global option.
.Pp
.Sy For example:
.Bd -literal -offset Ds
[xoreos]
width=1024
height=768
fullscreen=false

[nwn]
path=/home/drmccoy/games/nwn/
fullscreen=true
volume_music=0.500000
volume_sfx=0.850000
volume_voice=0.850000
.Ed
.Pp
You can then start the game with the target
.Dq nwn
with
.Pp
.Dl $ xoreos nwn
.Pp
and
.Nm
will do the rest.
This will start the game in the path
.Pa /home/drmccoy/games/nwn/ ,
running it in fullscreen at 1024\(mu768.
.Pp
As you can see with the volume options,
.Nm
also saves settings you made in the game' actual GUI into the
target's section of the config file.
.Pp
The place where
.Nm
expects the config file depends on your operating system:
.Bl -bullet -compact
.It
On GNU/Linux, the place is
.Pa $XDG_CONFIG_HOME/xoreos/xoreos.conf .
.Ev XDG_CONFIG_HOME
defaults to
.Pa $HOME/.config/ .
.It
On Mac OS X, the place is
.Pa $HOME/Library/Preferences/xoreos/xoreos.conf .
.It
On Windows,
.Pa xoreos.conf
is in the subdirectory
.Pa xoreos
in either
.Ev $APPDATA
or
.Ev $USERPROFILE ,
depending on your Windows version.
.El
.Sh OPTIONS
.Bl -tag -width Ds -compact
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl c Ar file
.It Fl Fl config= Ns Ar file
Load the config from file
.Ar file .
.It Fl p Ar dir
.It Fl Fl path= Ns Ar dir
Override the game path with
.Ar dir .
.It Fl w Ar size
.It Fl Fl width= Ns Ar size
Set the window's width to
.Ar size .
.It Fl h Ar size
.It Fl Fl height= Ns Ar size
Set the window's height to
.Ar size .
.It Fl f Ar bool
.It Fl Fl fullscreen= Ns Ar bool
Switch fullscreen on/off.
.It Fl k Ar bool
.It Fl Fl skipvideos= Ns Ar bool
Disable videos on/off.
.It Fl v Ar vol
.It Fl Fl volume= Ns Ar vol
Set global volume to
.Ar vol .
.It Fl m Ar vol
.It Fl Fl volume_music= Ns Ar vol
Set music volume to
.Ar vol .
.It Fl s Ar vol
.It Fl Fl volume_sfx= Ns Ar vol
Set SFX volume to
.Ar vol .
.It Fl o Ar vol
.It Fl Fl volume_voice= Ns Ar vol
Set voice volume to
.Ar vol .
.It Fl i Ar vol
.It Fl Fl volume_video= Ns Ar vol
Set video volume to
.Ar vol .
.It Fl q Ar lang
.It Fl Fl lang= Ns Ar lang
Set the game's language.
.It Fl Fl langtext= Ns Ar lang
Set the game's text language.
.It Fl Fl langvoice= Ns Ar lang
Set the game's voice language.
.It Fl d Ar dlvl
.It Fl Fl debug= Ns Ar dlvl
Set the debug channel verbosities to
.Ar dlvl .
.It Fl Fl debuggl= Ns Ar bool
Create OpenGL debug context.
.It Fl Fl listdebug
List all available debug channels.
.It Fl Fl listlangs
List all available languages for this target.
.It Fl Fl logfile= Ns Ar file
Write all debug output into this file too.
.It Fl Fl nologfile= Ns Ar bool
Don't write a log file.
.It Fl Fl consolelog= Ns Ar file
Write all debug console output into this file too.
.It Fl Fl noconsolelog= Ns Ar bool
Don't write a debug console log file.
.El
.Bl -tag -width Ds
.It Ar file
Absolute or relative path to a file.
.It Ar dir
Absolute or relative path to a directory.
.It Ar size
A positive integer.
.It Ar bool
.Dq true ,
.Dq yes ,
.Dq y ,
.Dq on ,
and
.Dq 1
are true, everything else is false.
.It Ar vol
A double ranging from 0.0 (min) \(en 1.0 (max).
.It Ar lang
A language identifier.
Full name, ISO 639-1 or ISO 639-2 language code;
or IETF language tag with ISO 639-1 and ISO 3166-1 country code.
Examples: en, de_de, hun, Czech, zh-tw, zh_cn, zh-cht, zh-chs.
.It Ar dlvl
A comma-separated list of a debug channel name, a colon and the desired
verbosity level. For example: GGraphics:4,GVideo:1,EEvents:9.
Use
.Dq All
to refer to all debug channels.
OpenGL debug channels need the debuggl option to be enabled and require
OpenGL driver support of the GL_ARB_debug_output extension.
.El
.Pp
Long-form command line option, like
.Fl Fl skipvideos ,
map directly to config options.
Options given on the command line will override any
options found in the config file for this session,
but will not save back to that config file.
.Sh EXAMPLES
.Dl $ xoreos -p/path/to/nwn/
.Pp
.Nm
will start the game in
.Pa /path/to/nwn/ .
Should a target with this path not yet exist in the config file,
.Nm
will create one named
.Dq nwn .
.Pp
.Dl $ xoreos -p/path/to/nwn/ foobar
.Pp
.Nm
will start the game in
.Pa /path/to/nwn/ .
If a target
.Dq foobar
does not yet exist in the config file,
.Nm
will create it.
.Pp
.Dl $ xoreos nwn
.Pp
.Nm
will start the game specified by target
.Dq nwn ,
which must exist in the config file already.
.Sh SEE ALSO
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
and
.Lk https://wiki.xoreos.org/ "its wiki"
.Ns .
.Sh AUTHORS
This program is part of the xoreos projects, and was written by
the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS.md
Changes in xoreos version 0.0.6
===============================

The tutorial in Knights of the Old Republic is now partially playable: the PC
can collect their belongings, put on their gear, let Trask open the first door,
open the second door and engage in mock combat.

Knights of the Old Republic II now lets you skip the prologue, showing off a bit
of the intro.

Jade Empire now has background music and a preliminary character generation
menu.

And as always, this release includes a ton of code quality changes, bug fixes
and initial implementations that are user-invisible for now. These include:
- Partial implementations of ActionScript and Scaleform GFx for Dragon Age menus
- Partial implementations of XACT WaveBanks and SoundBanks for Jade Empire
- Partial implementations of FMOD SampleBanks and Events for Dragon Age: Origins
- Partial implementations of Wwise SoundBanks for Dragon Age II
- WebM (Matroska + VPx) support for the Enhanced Edition of Neverwinter Nights
- Support for big-endian GFF4 files found in console ports
- Support for resource files found in mobile ports
- Support for more Neverwinter Nights 2 structures

In an attempt to modernize a bit, xoreos now requires a C++11-capable compiler.
This should hopefully not be a huge problem.


Changes in xoreos version 0.0.5
===============================

The two Knights of the old Republic games gained animation support, partially
working character creators, ingame menus, PC walking with walkmesh and trigger
evaluation, and preliminary conversations. They are now the most implemented
games in xoreos, overtaking Neverwinter Nights.

Neverwinter Nights received a fix in the animation system. Animations are now
correctly scaled with the height of the creature. Heads don't fly off of elves
anymore when they yawn, for example. The character creator was also partially
implemented.

Jade Empire now has an initial implementation for its main menu and lightmap
textures are correctly loaded.

And as always, this release includes a ton of code quality changes, bug fixes
and initial implementations that are user-invisible for now.


Changes in xoreos version 0.0.4
===============================

Neverwinter Nights now shows speech bubbles for conversation one-liners, as
used for cutscenes, bark strings and short NPC dialogues. Additionally, the
premium modules BioWare sold for Neverwinter Nights, including the three that
come with the Diamond Edition, can now be properly loaded and started.

An oversight in the handling of the texture fonts used in Neverwinter Nights
and the two Knights of Old Republic games has been fixed. This oversight
broke rendering of certain characters, most prominently of those used in
eastern European languages and the "smart" single quotation mark that's used
instead of an apostrophe in some strings found in the French versions.

Moreover, this release includes a lot of user-invisible code documentation
and quality fixes.


Changes in xoreos version 0.0.3
===============================

All targeted games now feature a working script system, with game scripts
being fired for the start of a campaign or module, when entering and leaving
areas, and when clicking on in-game object. The singular exception is the
Nintendo DS game Sonic Chronicles: The Dark Brotherhood, which doesn't seem to
feature any scripts at all.

The vast majority of engine functions, the functions that are called by the
scripts and that do the actual work of tracking and changing the game state,
are still missing, though. Per game there are about 850 functions (with some
overlap) that need to be implemented. We currently have about 90, per game, of
these written and working within xoreos. Moreover, many of the functions still
missing depend on features not yet implemented.

Apart from the script system changes, 0.0.3 also comes with support for
reflective environment mapping in Neverwinter Nights and the two Knights of
the Old Republic games. The "metallic" armor and area parts that were rendered
rendered transparent in xoreos are now properly reflective. This can be seen,
for example, in the Sith troopers in Knights of the Old Republic, in various
plate armor worn by NPC in Neverwinter Nights, as well as the metallic floors
on the planet of Taris and the icy wastes of Cania. For Neverwinter Nights,
xoreos now also correctly smoothes the vertex normals of (binary) models, so
that the metallic effect is not broken by sharp polygon edges.


Changes in xoreos version 0.0.2
===============================

This is the first official release of xoreos, together with xoreos-tools.

In this first release, all targeted games work insofar as that they at least
show basic in-game areas. I.e. you can start the game, xoreos loads the game
resources, loads a campaign or module, and then shows an area of the game.

Within the area, you can fly around in a "spectator" mode, using the common
first-person WASD control scheme. Moving the mouse while holding down the
middle mouse button rotates the camera. With Ctrl+D, a debug console drops
down, allowing for general resource dumping and the loading of different
areas, modules and/or campaigns.

A few games, specifically Neverwinter Nights and Knights of the Old Republic,
also show a main menu, although the latter's is not as extensive yet. The
former also shows a few in-game menu elements.

Additionally, Neverwinter Nights also has a script system hooked up, and
preliminary dialogue support. This means that clicking on an NPC opens up
its conversation dialog, and some of the script commands will be executed.
For example, the door in the first area of the original campaign's prelude
opens after speaking to Bim and telling him that no tutorial is necessary.
However, triggering the tutorial leads to the scripts looping endlessly,
because the necessary game functions are not implemented yet.

Further gameplay is still missing. At the moment, none of the other games
have a script system.


src/xoreos.cpp
/* xoreos - A reimplementation of BioWare's Aurora engine
 *
 * xoreos is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  The project's main entry point.
 */

#define SDL_MAIN_HANDLED

#include <cassert>
#include <cstdio>

#include <vector>

#include "src/cline.h"
#include "src/engines.h"

#include "src/common/ustring.h"
#include "src/common/util.h"
#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/filepath.h"
#include "src/common/threads.h"
#include "src/common/debugman.h"
#include "src/common/configman.h"
#include "src/common/random.h"
#ifdef ENABLE_XML
#include "src/common/xml.h"
#endif

#include "src/aurora/resman.h"
#include "src/aurora/2dareg.h"
#include "src/aurora/language.h"
#include "src/aurora/talkman.h"
#include "src/aurora/util.h"

#include "src/aurora/nwscript/objectman.h"
#include "src/aurora/nwscript/functionman.h"

#include "src/graphics/queueman.h"
#include "src/graphics/graphics.h"
#include "src/graphics/camera.h"
#include "src/graphics/windowman.h"
#include "src/graphics/yuv_to_rgb.h"

#include "src/graphics/aurora/textureman.h"
#include "src/graphics/aurora/cursorman.h"
#include "src/graphics/aurora/fontman.h"

#include "src/graphics/mesh/meshman.h"

#include "src/graphics/shader/surfaceman.h"
#include "src/graphics/shader/materialman.h"
#include "src/graphics/shader/shader.h"

#include "src/sound/sound.h"

#include "src/events/requests.h"
#include "src/events/events.h"
#include "src/events/timerman.h"
#include "src/events/notifications.h"

#include "src/engines/enginemanager.h"
#include "src/engines/gamethread.h"

#include "src/engines/aurora/tokenman.h"
#include "src/engines/aurora/flycamera.h"

static void initPlatform();
static void initConfig();

static void init();
static void deinit();

static void listDebug();

static bool configFileIsBroken = false;

int main(int argc, char **argv) {
	initPlatform();
	initConfig();

	std::vector<Common::UString> args;
	Common::UString target;

	try {
		Common::Platform::getParameters(argc, argv, args);

		int code;
		if (!parseCommandline(args, target, code))
			return code;

	} catch (...) {
		Common::exceptionDispatcherError();

		return 1;
	}

	// Check the requested target
	if (target.empty() || !ConfigMan.hasGame(target)) {
		Common::UString path = ConfigMan.getString("path", "");
		if (path.empty()) {
			if (ConfigMan.getBool("listdebug", false)) {
				listDebug();
				return 0;
			}

			error("Neither an existing target, nor a path specified");
		}

		bool useExisting = false;
		if (target.empty()) {
			target = ConfigMan.findGame(path);
			if (!target.empty()) {
				warning("No target specified, but found a target with a matching path");
				useExisting = true;
			}
		}

		target = ConfigMan.createGame(path, target);
		if (target.empty())
			error("Failed creating a new config target for the game");

		if (!useExisting)
			warning("Creating a new target for this game");
	}

	if (!ConfigMan.setGame(target) || !ConfigMan.isInGame())
		error("No target \"%s\" in the config file", target.c_str());

	/* Open the log file.
	 *
	 * NOTE: A log is opened by default, unless the logfile config value
	 *       is set to an empty string or nologfile is set to true.
	 */
	Common::UString logFile = DebugMan.getDefaultLogFile();
	if (ConfigMan.hasKey("logfile"))
		logFile = ConfigMan.getString("logfile", "");
	if (ConfigMan.getBool("nologfile", false))
		logFile.clear();

	if (!logFile.empty())
		if (!DebugMan.openLogFile(logFile))
			warning("Failed to open log file \"%s\" for writing", logFile.c_str());

	DebugMan.logCommandLine(args);

	status("Target \"%s\"", target.c_str());

	Common::UString dirArg = ConfigMan.getString("path", "");
	if (dirArg.empty())
		error("Target \"%s\" is missing a path", target.c_str());

	Common::UString baseDir;
	try {
		baseDir = Common::FilePath::canonicalize(dirArg);
	} catch (...) {
		error("Invalid path \"%s\"", dirArg.c_str());
	}

	if (!Common::FilePath::isDirectory(baseDir) && !Common::FilePath::isRegularFile(baseDir))
		error("No such file or directory \"%s\"", baseDir.c_str());

	std::list<const Engines::EngineProbe *> probes;

	Engines::GameThread *gameThread = new Engines::GameThread;
	try {
		// Enable requested debug channels
		DebugMan.setVerbosityLevelsFromConfig();

		// Initialize all necessary subsystems
		init();

		// Create all our engine probes
		createEngineProbes(probes);

		// Probe and create the game engine
		gameThread->init(baseDir, probes);

		if (ConfigMan.getBool("listdebug", false)) {
			// List debug channels
			listDebug();
		} else {
			// Run the game
			gameThread->run();
			EventMan.runMainLoop();
		}

	} catch (...) {
		Common::exceptionDispatcherError();
	}

	if (EventMan.fatalErrorRaised())
		std::exit(1);

	status("Shutting down");

	try {
		delete gameThread;
	} catch (...) {
	}

	destroyEngineProbes(probes);

	try {
		// Sync changed debug channel settings
		DebugMan.setConfigToVerbosityLevels();

		// Configs changed, we should save them
		if (ConfigMan.changed()) {
			// But don't clobber a broken save
			if (!configFileIsBroken)
				ConfigMan.save();
		}
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	deinit();
	return 0;
}

static void initPlatform() {
	try {
		Common::Platform::init();
	} catch (...) {
		Common::exceptionDispatcherError("Failed to initialize the low-level platform-specific subsytem");
		std::exit(1);
	}
}

static void initConfig() {
	bool newConfig = false;
	if (!ConfigMan.load()) {
		// Loading failed, create an empty config file

		ConfigMan.create();

		// Mark the config as either broken or new
		if (ConfigMan.fileExists())
			configFileIsBroken = true;
		else
			newConfig = true;
	}

	ConfigMan.setInt   (Common::kConfigRealmDefault, "width" ,     800);
	ConfigMan.setInt   (Common::kConfigRealmDefault, "height",     600);
	ConfigMan.setBool  (Common::kConfigRealmDefault, "fullscreen", false);
	ConfigMan.setInt   (Common::kConfigRealmDefault, "fsaa",       0);

	ConfigMan.setDouble(Common::kConfigRealmDefault, "volume"      , 1.0);
	ConfigMan.setDouble(Common::kConfigRealmDefault, "volume_music", 1.0);
	ConfigMan.setDouble(Common::kConfigRealmDefault, "volume_sfx"  , 1.0);
	ConfigMan.setDouble(Common::kConfigRealmDefault, "volume_voice", 1.0);
	ConfigMan.setDouble(Common::kConfigRealmDefault, "volume_video", 1.0);

	ConfigMan.setBool(Common::kConfigRealmDefault, "showfps", false);

	ConfigMan.setBool(Common::kConfigRealmDefault, "skipvideos", false);

	ConfigMan.setBool(Common::kConfigRealmDefault, "saveconf", true);

	// Populate the new config with the defaults
	if (newConfig) {
		ConfigMan.setDefaults();
		ConfigMan.save();
	}
}

static void listDebug() {
	std::vector<Common::UString> names, descriptions;
	DebugMan.getDebugChannels(names, descriptions);

	assert(names.size() == descriptions.size());

	size_t maxNameLength = 0;
	for (std::vector<Common::UString>::const_iterator n = names.begin(); n != names.end(); n++)
		maxNameLength = MAX(maxNameLength, n->size());

	assert(maxNameLength < INT_MAX);

	for (size_t i = 0; i < names.size(); i++)
		std::printf("%-*s - %s\n", (int) maxNameLength, names[i].c_str(), descriptions[i].c_str());
}

static void init() {
	// Init threading system
	Common::initThreads();

#ifdef ENABLE_XML
	// Init libxml2
	Common::initXML();
#endif

	// Init subsystems
	GfxMan.init();
	status("Graphics subsystem initialized");
	SoundMan.init();
	status("Sound subsystem initialized");
	EventMan.init();
	status("Event subsystem initialized");
}

static void deinit() {
	// Deinit subsystems
	try {
		if (Common::initedThreads()) {
			EventMan.deinit();
			SoundMan.deinit();
			GfxMan.deinit();
		}
	} catch (...) {
	}

#ifdef ENABLE_XML
	// Deinit libxml2
	Common::deinitXML();
#endif

	// Destroy global singletons
	Graphics::Aurora::FontManager::destroy();
	Graphics::Aurora::CursorManager::destroy();
	Graphics::Aurora::TextureManager::destroy();

	Aurora::LanguageManager::destroy();
	Aurora::TalkManager::destroy();
	Aurora::TwoDARegistry::destroy();
	Aurora::ResourceManager::destroy();
	Aurora::FileTypeManager::destroy();

	Aurora::NWScript::ObjectManager::destroy();
	Aurora::NWScript::FunctionManager::destroy();

	Engines::EngineManager::destroy();
	Engines::TokenManager::destroy();
	Engines::FlyCamera::destroy();

	Events::EventsManager::destroy();
	Events::RequestManager::destroy();
	Events::TimerManager::destroy();

	Sound::SoundManager::destroy();

	Graphics::GraphicsManager::destroy();
	Graphics::QueueManager::destroy();
	Graphics::CameraManager::destroy();
	Graphics::WindowManager::destroy();
	Graphics::YUVToRGBManager::destroy();

	Graphics::Mesh::MeshManager::destroy();

	Graphics::Shader::SurfaceManager::destroy();
	Graphics::Shader::MaterialManager::destroy();
	Graphics::Shader::ShaderManager::destroy();

	Events::NotificationManager::destroy();

	Common::DebugManager::destroy();
	Common::ConfigManager::destroy();
	Common::Random::destroy();
}
